library/xml-model/src/main/java/org/cru/godtools/xml/model/Manifest.kt

```diff
@@ -205,88 +220,85 @@ class Manifest : BaseModel, Styles {
 
-        return buildList {
-            while (parser.next() != XmlPullParser.END_TAG) {
-                if (parser.eventType != XmlPullParser.START_TAG) continue
+        while (next() != XmlPullParser.END_TAG) {
+            if (eventType != XmlPullParser.START_TAG) continue
 
-                when (parser.namespace) {
-                    XMLNS_MANIFEST -> when (parser.name) {
-                        Category.XML_CATEGORY -> add(Category(this@Manifest, parser))
-                        else -> parser.skipTag()
-                    }
-                    else -> parser.skipTag()
+            when (namespace) {
+                XMLNS_MANIFEST -> when (name) {
+                    Category.XML_CATEGORY -> add(Category(this@Manifest, this@parseCategories))
+                    else -> skipTag()
                 }
+                else -> skipTag()
             }
         }
     }
 
     private class PagesData {
-        var tractPages: List<TractPage>? = null
         val aemImports = mutableListOf<Uri>()
+        val tractPageTasks = mutableListOf<Deferred<TractPage>>()
     }
 
     @WorkerThread
-    private fun parsePages(parser: XmlPullParser, parseFile: suspend (String) -> CloseableXmlPullParser) = runBlocking {
-        parser.require(XmlPullParser.START_TAG, XMLNS_MANIFEST, XML_PAGES)
+    private fun XmlPullParser.parsePages(
+        scope: CoroutineScope,
+        parseFile: suspend (String) -> CloseableXmlPullParser
+    ) = PagesData().also { result ->
+        require(XmlPullParser.START_TAG, XMLNS_MANIFEST, XML_PAGES)
 
         // process any child elements
-        val result = PagesData()
-        result.tractPages = buildList {
-            while (parser.next() != XmlPullParser.END_TAG) {
-                if (parser.eventType != XmlPullParser.START_TAG) continue
-
-                when (parser.namespace) {
-                    XMLNS_MANIFEST -> when (parser.name) {
-                        XML_PAGES_PAGE -> {
-                            val fileName = parser.getAttributeValue(null, XML_PAGES_PAGE_FILENAME)
-                            val srcFile = parser.getAttributeValue(null, XML_PAGES_PAGE_SRC)
-                            val pos = size
-                            parser.skipTag()
-
-                            if (srcFile != null)
-                                add(async { parseFile(srcFile).use { TractPage(this@Manifest, pos, fileName, it) } })
+        while (next() != XmlPullParser.END_TAG) {
+            if (eventType != XmlPullParser.START_TAG) continue
+
+            when (namespace) {
+                XMLNS_MANIFEST -> when (name) {
+                    XML_PAGES_PAGE -> {
+                        val fileName = getAttributeValue(null, XML_PAGES_PAGE_FILENAME)
+                        val src = getAttributeValue(null, XML_PAGES_PAGE_SRC)
+                        skipTag()
+
+                        if (src != null) {
+                            val pos = result.tractPageTasks.size
```

Comment:
It doesn't need to be within the null check, but it will make more sense when I add on parsing of lesson pages. lesson pages will be stored in a separate list, so `pos` is determined based on which type of pages are being parsed.

see: https://github.com/CruGlobal/godtools-android/pull/1499/files#diff-c131b629ab38d425ee903647d630411276d74260f262e23e4570c519fb1ce045L263-R287